Refactor FeatureModelParser, add unittests and Location

s9latimm opened the pull request.

s9latimm committed:
Add location

s9latimm committed:
Fix dummy test

vulder commented on include/vara/Feature/Feature.h:
```diff
@@ -34,9 +35,11 @@ class Feature {
   FeatureListType Alternatives;
   RelationshipListTy Relationships;
   bool Opt;
+  std::optional<std::unique_ptr<Location>> Loc;
```
We don't need to put the location on the heap, especially not when we copy it anyhow when calling `getLocation`

vulder commented on include/vara/Feature/Feature.h:
```diff
@@ -105,14 +108,23 @@ class Feature {
   }
 
+  [[nodiscard]] std::optional<std::unique_ptr<Location>> getLocation() const {
+    return Loc ? std::optional{std::make_unique<Location>(
+                     Loc->get()->getPath(), Loc->get()->getStart(),
+                     Loc->get()->getEnd())}
+               : std::nullopt;
```
Just return a 'std::optional<Location>'
```
return Loc;
```

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -21,17 +21,22 @@ class FeatureModel {
 
 private:
   string Name;
+  string Path;
```
```suggestion
  std::filesystem::path RootPath;
```
This makes it more explicit what the intend behind the path is.
PS: we could also use `std::filesystem::path` for paths, making the path platform independent and giving us methods to interact with the path.

vulder commented on include/vara/Feature/Location.h:
```diff
@@ -0,0 +1,46 @@
+class Location {
+private:
+  std::string Path;
+  std::optional<std::pair<int, int>> Start;
```
```suggestion
  std::optional<HelperClassFoo> Start;
```

```
class HelperClassFoo : private std::pair<int, int> {
  int getLineNumber() { ... }
  ...
}
```

We should add a nested class here to have a type with a name and accessor methods that have meaning full names, otherwise, the code is quite hard to understand.

vulder commented on include/vara/Feature/XmlParser.h:
```diff
@@ -34,6 +35,7 @@ class XmlParser {
   void parseOptions(xmlNode *N, bool Num);
   void parseConstraints(xmlNode *N);
   void parseVm(xmlNode *N);
+  static std::pair<int, int> parseLocation(xmlNode *N);
```
Hmm, is there a reason why you made this one static and the other methods are not static?

vulder commented on lib/Feature/XmlParser.cpp:
```diff
@@ -30,13 +31,37 @@ void XmlParser::parseConfigurationOption(xmlNode *N, bool Num = false) {
+      } else if (!xmlStrcmp(Head->name,
+                            reinterpret_cast<constXmlCharPtr>("location"))) {
+        std::string Path;
+        std::optional<std::pair<int, int>> Start = std::nullopt;
+        std::optional<std::pair<int, int>> End = std::nullopt;
```
```suggestion
        std::optional<std::pair<int, int>> Start;
        std::optional<std::pair<int, int>> End;
```
Default construction already gives you an empty optional.

vulder commented on include/vara/Feature/Feature.h:
```diff
@@ -105,14 +108,23 @@ class Feature {
 public:
-  BinaryFeature(string Name, bool Opt) : Feature(std::move(Name), Opt) {}
+  BinaryFeature(string Name, bool Opt,
+                std::optional<std::unique_ptr<Location>> Loc)
```
I think we could default construct the optional or create a second ctor that takes no location. This would make it easier for callers because they do not have to pass in std::nullopts if the location is missing.

vulder commented on unittests/Feature/Feature.cpp:
```diff
@@ -4,7 +4,7 @@
 namespace vara::feature {
 TEST(BinaryFeature, getName) {
-  BinaryFeature F("Test", false);
+  BinaryFeature F("Test", false, std::nullopt);
```
```suggestion
  BinaryFeature F("Test", false);
```
Could be remove when adapting the ctor behavior.

vulder reviewed (changes requested):
Looks good overall. Found just a few things where we could improve the design.

s9latimm committed:
Improved design

s9latimm committed:
Replace unnecessary casting

s9latimm committed:
Some tests

s9latimm committed:
only latest stage

vulder commented on .travis.yml:
```diff
@@ -12,25 +12,25 @@ jobs:
     ################
     # Clang on linux
 
```
Clearly this part needs to be fixed :)

vulder commented on include/vara/Feature/Feature.h:
```diff
@@ -105,14 +108,22 @@ class Feature {
+  [[nodiscard]] std::optional<Location> getLocation() const {
+    return Loc ? std::optional{Location(Loc->getPath(), Loc->getStart(),
+                                        Loc->getEnd())}
+               : std::nullopt;
```
```suggestion
    return Loc;
```
Just return the optional, a copy will be made automatically.

vulder commented on include/vara/Feature/Location.h:
```diff
@@ -0,0 +1,52 @@
+
+class Location {
+public:
+  class TableEntry : private std::pair<int, int> {
```
`TableEntry` sounds a bit generic, maybe something like `LineAndCharacterOffset` or a bit shorter `LineCharOffset` or `LineColumnOffset`
As I understand this the two values are line number and column/char number.

vulder commented on include/vara/Feature/Location.h:
```diff
@@ -0,0 +1,52 @@
+  public:
+    TableEntry(int Line, int Column) : std::pair<int, int>(Line, Column) {}
+
+    [[nodiscard]] int getLine() const { return this->first; }
```
```suggestion
    [[nodiscard]] int getLineNumber() const { return this->first; }
```

vulder commented on include/vara/Feature/Location.h:
```diff
@@ -0,0 +1,52 @@
+
+    [[nodiscard]] int getLine() const { return this->first; }
+
+    [[nodiscard]] int getColumn() const { return this->second; }
```
```suggestion
    [[nodiscard]] int getColumnOffset() const { return this->second; }
```

vulder commented on lib/Feature/XmlParser.cpp:
```diff
@@ -62,24 +77,26 @@ void XmlParser::parseConfigurationOption(xmlNode *N, bool Num = false) {
     } else {
       Features.try_emplace(
           Name, std::make_unique<NumericFeature>(
                     Name, Opt,
-                    std::variant<std::pair<int, int>, std::vector<int>>(Vals)));
+                    std::variant<std::pair<int, int>, std::vector<int>>(Vals),
+                    std::move(Loc)));
     }
   } else {
-    Features.try_emplace(Name, std::make_unique<BinaryFeature>(Name, Opt));
+    Features.try_emplace(
```
The try_emplace could 'fail' and not insert the new element if the key is already present. Maybe we should check the return value with an assert.
```
auto emplaced = Features.try_emplace(...)
assert(emplaced.second && "Feature value could not be inserted, values was already present.");
```

vulder commented on lib/Feature/XmlParser.cpp:
```diff
@@ -113,34 +129,60 @@ void XmlParser::parseConstraints(xmlNode *N) {
   if (Features.find("root") == Features.end()) {
-    Features.try_emplace("root",
-                         std::make_unique<BinaryFeature>("root", false));
+    Features.try_emplace(
```
same here

vulder commented on tools/fm-viewer/FeatureModelViewer.cpp:
```diff
@@ -1,12 +1,14 @@
+    auto P = vara::feature::XmlParser(argv[1], argv[2]);
+    if (P.parse()) {
+      std::unique_ptr<vara::feature::FeatureModel> FM = P.buildFeatureModel();
+      //      FM->dump();
```
```suggestion
```

vulder reviewed (changes requested):
Looks good, found just 2 small nits.
After those and the travis fix we can merge this. Then I'll start working on the python bindings.

s9latimm committed:
Refactor parser

M	include/vara/Feature/Feature.h
M	include/vara/Feature/Location.h
M	lib/Feature/CMakeLists.txt
M	tools/fm-viewer/FeatureModelViewer.cpp
A	unittests/Feature/FeatureModelParser.cpp

s9latimm committed:
Static DTD

M	include/vara/Feature/FeatureModelParser.h
M	lib/Feature/FeatureModelParser.cpp
M	tools/fm-viewer/FeatureModelViewer.cpp
M	unittests/Feature/CMakeLists.txt
M	unittests/Feature/FeatureModelParser.cpp

s9latimm committed:
Merge branch 'f-Tests' into f-Location

M	include/vara/Feature/Feature.h
A	unittests/Feature/BinaryFeature.cpp
M	unittests/Feature/CMakeLists.txt
D	unittests/Feature/Feature.cpp
A	unittests/Feature/NumericFeature.cpp

s9latimm committed:
More tests

M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/Location.h
M	lib/Feature/FeatureModelParser.cpp
M	unittests/Feature/CMakeLists.txt
M	unittests/Feature/FeatureModelParser.cpp
A	unittests/Feature/Location.cpp
M	unittests/Feature/NumericFeature.cpp

s9latimm committed:
Fix travis

M	.travis.yml

s9latimm committed:
Fix tidy

M	unittests/Feature/FeatureModelParser.cpp

s9latimm committed:
Merge branch 'vara-dev' into f-Location

s9latimm committed:
Minimal travis

M	.travis.yml

s9latimm committed:
Merge branch 'f-Location' of github.com:se-passau/vara-feature into f-Location

M	.gitignore
M	.gitmodules
M	.travis.yml
M	CMakeLists.txt
M	README.md
A	bindings/CMakeLists.txt
A	bindings/python/CMakeLists.txt
A	bindings/python/tests/CMakeLists.txt
A	bindings/python/tests/TEST_INPUTS/example_feature_model.xml
A	bindings/python/tests/TEST_INPUTS/vm.dtd
A	bindings/python/tests/__init__.py
A	bindings/python/tests/feature/__init__.py
A	bindings/python/tests/feature/test_feature.py
A	bindings/python/tests/feature_model/__init__.py
A	bindings/python/tests/feature_model/test_feature_model.py
A	bindings/python/vara-feature/CMakeLists.txt
A	bindings/python/vara-feature/pybind_Feature.cpp
A	bindings/python/vara-feature/pybind_FeatureModel.cpp
A	bindings/python/vara-feature/pybind_Init.cpp
A	bindings/python/vara-feature/pybind_XmlParser.cpp
A	bindings/python/vara-feature/pybind_common.h
A	bindings/python/vara-feature/pybind_llvm.cpp
A	external/pybind11
M	include/vara/Feature/Feature.h
A	setup.py

s9latimm committed:
Fix tidy

M	unittests/Feature/BinaryFeature.cpp
M	unittests/Feature/FeatureModelParser.cpp

s9latimm committed:
Fix pybind

M	bindings/python/vara-feature/CMakeLists.txt
A	bindings/python/vara-feature/pybind_FeatureModelXmlParser.cpp
D	bindings/python/vara-feature/pybind_XmlParser.cpp

s9latimm committed:
Disable test_feature_model

M	bindings/python/tests/feature_model/test_feature_model.py
M	unittests/Feature/NumericFeature.cpp

vulder commented on include/vara/Feature/Feature.h:
```diff
@@ -139,6 +135,8 @@ class NumericFeature : public Feature {
                  std::optional<Location> Loc = std::nullopt)
       : Feature(std::move(Name), Opt, std::move(Loc)), Vals(std::move(Vals)) {}
 
+  std::variant<std::pair<int, int>, std::vector<int>> getVals() { return Vals; }
```
Should we add a using decl for this?

vulder commented on unittests/Feature/BinaryFeature.cpp:
```diff
@@ -0,0 +1,63 @@
+
+namespace vara::feature {
+TEST(BinaryFeature, BinaryFeatureBasics) {
+  BinaryFeature A("A", true);
```
```suggestion
  BinaryFeature A("A", true);

```

vulder commented on unittests/Feature/BinaryFeature.cpp:
```diff
@@ -0,0 +1,63 @@
+
+TEST(BinaryFeature, BinaryFeatureRoot) {
+  BinaryFeature A("A", false);
+  BinaryFeature B("B", false);
```
```suggestion
  BinaryFeature B("B", false);

```

vulder commented on unittests/Feature/BinaryFeature.cpp:
```diff
@@ -0,0 +1,63 @@
+TEST(BinaryFeature, BinaryFeatureRoot) {
+  BinaryFeature A("A", false);
+  BinaryFeature B("B", false);
+  B.addParent(&A);
```
```suggestion
  B.addParent(&A);

```

vulder commented on unittests/Feature/BinaryFeature.cpp:
```diff
@@ -0,0 +1,63 @@
+
+TEST(BinaryFeature, BinaryFeatureChildren) {
+  BinaryFeature A("A", false);
+  BinaryFeature B("B", false);
```
```suggestion
  BinaryFeature B("B", false);

```

vulder commented on unittests/Feature/BinaryFeature.cpp:
```diff
@@ -0,0 +1,63 @@
+TEST(BinaryFeature, BinaryFeatureChildren) {
+  BinaryFeature A("A", false);
+  BinaryFeature B("B", false);
+  B.addChild(&A);
```
```suggestion
  B.addChild(&A);

```

vulder commented on unittests/Feature/BinaryFeature.cpp:
```diff
@@ -0,0 +1,63 @@
+
+TEST(BinaryFeature, BinaryFeatureExclude) {
+  BinaryFeature A("A", false);
+  BinaryFeature B("B", false);
```
```suggestion
  BinaryFeature B("B", false);

```

vulder commented on unittests/Feature/BinaryFeature.cpp:
```diff
@@ -0,0 +1,63 @@
+TEST(BinaryFeature, BinaryFeatureExclude) {
+  BinaryFeature A("A", false);
+  BinaryFeature B("B", false);
+  B.addExclude(&A);
```
```suggestion
  B.addExclude(&A);

```

vulder commented on unittests/Feature/BinaryFeature.cpp:
```diff
@@ -0,0 +1,63 @@
+
+TEST(BinaryFeature, BinaryFeatureImplications) {
+  BinaryFeature A("A", false);
+  BinaryFeature B("B", false);
```
```suggestion
  BinaryFeature B("B", false);

```

vulder commented on unittests/Feature/BinaryFeature.cpp:
```diff
@@ -0,0 +1,63 @@
+TEST(BinaryFeature, BinaryFeatureImplications) {
+  BinaryFeature A("A", false);
+  BinaryFeature B("B", false);
+  B.addImplication(&A);
```
```suggestion
  B.addImplication(&A);

```

vulder commented on unittests/Feature/BinaryFeature.cpp:
```diff
@@ -0,0 +1,63 @@
+
+TEST(BinaryFeature, BinaryFeatureAlternatives) {
+  BinaryFeature A("A", false);
+  BinaryFeature B("B", false);
```
```suggestion
  BinaryFeature B("B", false);

```

vulder commented on unittests/Feature/BinaryFeature.cpp:
```diff
@@ -0,0 +1,63 @@
+TEST(BinaryFeature, BinaryFeatureAlternatives) {
+  BinaryFeature A("A", false);
+  BinaryFeature B("B", false);
+  B.addAlternative(&A);
```
```suggestion
  B.addAlternative(&A);

```

vulder commented on unittests/Feature/FeatureModelParser.cpp:
```diff
@@ -0,0 +1,269 @@
+      vara::feature::FeatureModelXmlParser(TestDoc).verifyFeatureModel());
+}
+
+TEST(FeatureModelXmlParser, BuildFeatureModel) {
```
This test is much to complicated, we need to split it up into separated tests that are easy to understand and only check a specific thing. We can talk about this in the next meeting.

vulder commented on unittests/Feature/Location.cpp:
```diff
@@ -0,0 +1,16 @@
+TEST(Location, LocationBasics) {
+  auto L = Location(std::filesystem::current_path(),
+                    Location::LineColumnOffset(1, 4),
+                    Location::LineColumnOffset(3, 5));
```
```suggestion
                    Location::LineColumnOffset(3, 5));

```

vulder commented on unittests/Feature/NumericFeature.cpp:
```diff
@@ -0,0 +1,78 @@
+
+namespace vara::feature {
+TEST(NumericFeature, NumericFeatureBasics) {
+  NumericFeature A("A", true, std::pair<int, int>(0, 1));
```
```suggestion
  NumericFeature A("A", true, std::pair<int, int>(0, 1));

```

vulder commented on unittests/Feature/NumericFeature.cpp:
```diff
@@ -0,0 +1,78 @@
+}
+
+TEST(NumericFeature, NumericFeaturePair) {
+  NumericFeature A("A", false, std::pair<int, int>(0, 1));
```
```suggestion
  NumericFeature A("A", false, std::pair<int, int>(0, 1));

```

vulder commented on unittests/Feature/NumericFeature.cpp:
```diff
@@ -0,0 +1,78 @@
+}
+
+TEST(NumericFeature, NumericFeatureVector) {
+  NumericFeature A("A", false, std::vector<int>{0, 1, 2, 3});
```
```suggestion
  NumericFeature A("A", false, std::vector<int>{0, 1, 2, 3});

```

vulder commented on unittests/Feature/NumericFeature.cpp:
```diff
@@ -0,0 +1,78 @@
+
+TEST(NumericFeature, NumericFeatureRoot) {
+  NumericFeature A("A", false, std::pair<int, int>(0, 1));
+  NumericFeature B("B", false, std::pair<int, int>(0, 1));
```
```suggestion
  NumericFeature B("B", false, std::pair<int, int>(0, 1));

```

vulder commented on unittests/Feature/NumericFeature.cpp:
```diff
@@ -0,0 +1,78 @@
+TEST(NumericFeature, NumericFeatureRoot) {
+  NumericFeature A("A", false, std::pair<int, int>(0, 1));
+  NumericFeature B("B", false, std::pair<int, int>(0, 1));
+  B.addParent(&A);
```
```suggestion
  B.addParent(&A);

```

vulder commented on unittests/Feature/NumericFeature.cpp:
```diff
@@ -0,0 +1,78 @@
+
+TEST(NumericFeature, NumericFeatureChildren) {
+  NumericFeature A("A", false, std::pair<int, int>(0, 1));
+  NumericFeature B("B", false, std::pair<int, int>(0, 1));
```
```suggestion
  NumericFeature B("B", false, std::pair<int, int>(0, 1));

```

vulder commented on unittests/Feature/NumericFeature.cpp:
```diff
@@ -0,0 +1,78 @@
+TEST(NumericFeature, NumericFeatureChildren) {
+  NumericFeature A("A", false, std::pair<int, int>(0, 1));
+  NumericFeature B("B", false, std::pair<int, int>(0, 1));
+  B.addChild(&A);
```
```suggestion
  B.addChild(&A);

```

vulder commented on unittests/Feature/NumericFeature.cpp:
```diff
@@ -0,0 +1,78 @@
+
+TEST(NumericFeature, NumericFeatureExclude) {
+  NumericFeature A("A", false, std::pair<int, int>(0, 1));
+  NumericFeature B("B", false, std::pair<int, int>(0, 1));
```
```suggestion
  NumericFeature B("B", false, std::pair<int, int>(0, 1));

```

vulder commented on unittests/Feature/NumericFeature.cpp:
```diff
@@ -0,0 +1,78 @@
+TEST(NumericFeature, NumericFeatureExclude) {
+  NumericFeature A("A", false, std::pair<int, int>(0, 1));
+  NumericFeature B("B", false, std::pair<int, int>(0, 1));
+  B.addExclude(&A);
```
```suggestion
  B.addExclude(&A);

```

vulder commented on unittests/Feature/NumericFeature.cpp:
```diff
@@ -0,0 +1,78 @@
+
+TEST(NumericFeature, NumericFeatureImplications) {
+  NumericFeature A("A", false, std::pair<int, int>(0, 1));
+  NumericFeature B("B", false, std::pair<int, int>(0, 1));
```
```suggestion
  NumericFeature B("B", false, std::pair<int, int>(0, 1));

```

vulder commented on unittests/Feature/NumericFeature.cpp:
```diff
@@ -0,0 +1,78 @@
+TEST(NumericFeature, NumericFeatureImplications) {
+  NumericFeature A("A", false, std::pair<int, int>(0, 1));
+  NumericFeature B("B", false, std::pair<int, int>(0, 1));
+  B.addImplication(&A);
```
```suggestion
  B.addImplication(&A);

```

vulder commented on unittests/Feature/NumericFeature.cpp:
```diff
@@ -0,0 +1,78 @@
+
+TEST(NumericFeature, NumericFeatureAlternatives) {
+  NumericFeature A("A", false, std::pair<int, int>(0, 1));
+  NumericFeature B("B", false, std::pair<int, int>(0, 1));
```
```suggestion
  NumericFeature B("B", false, std::pair<int, int>(0, 1));

```

vulder commented on unittests/Feature/NumericFeature.cpp:
```diff
@@ -0,0 +1,78 @@
+TEST(NumericFeature, NumericFeatureAlternatives) {
+  NumericFeature A("A", false, std::pair<int, int>(0, 1));
+  NumericFeature B("B", false, std::pair<int, int>(0, 1));
+  B.addAlternative(&A);
```
```suggestion
  B.addAlternative(&A);

```

s9latimm committed:
Test bindings

M	bindings/python/tests/feature_model/test_feature_model.py
M	include/vara/Feature/FeatureModel.h
M	lib/Feature/FeatureModelParser.cpp
M	unittests/Feature/FeatureModelParser.cpp

vulder commented on .travis.yml:
```diff
@@ -3,34 +3,33 @@ language: cpp
+            - clang-9
+            - llvm-9-dev
+            - clang-tidy-10
+            - libxml2-dev
```
Small nit: Should we put general deps into the `before_install`? This makes it easier to add new builds to the matrix.

vulder reviewed (changes requested):
Looks good :)
Most of my comments can just be batch applied via github and fix the test case alignment into a (given/when/then) style, separating the different stages.
Regarding, the travis nit and the big test, let's talk about it in the next 1:1.

s9latimm committed:
Fix argc

M	tools/fm-viewer/FeatureModelViewer.cpp

s9latimm committed:
Fix viewer

D	bindings/python/tests/TEST_INPUTS/vm.dtd
M	include/vara/Feature/Location.h
M	test/test.xml

s9latimm committed:
Add ws

M	include/vara/Feature/Feature.h
M	lib/Feature/Feature.cpp
M	unittests/Feature/BinaryFeature.cpp
M	unittests/Feature/Location.cpp
M	unittests/Feature/NumericFeature.cpp

s9latimm committed:
Run CI

M	.travis.yml

s9latimm committed:
Add libs

M	CMakeLists.txt
M	lib/Feature/CMakeLists.txt

s9latimm committed:
Build w/ clang

M	.travis.yml

s9latimm committed:
List ld

M	.travis.yml
M	CMakeLists.txt
M	lib/Feature/CMakeLists.txt

s9latimm committed:
Add some packages

M	.travis.yml

s9latimm committed:
Add some packages

M	.travis.yml

s9latimm committed:
Use experimental

M	.travis.yml
M	include/vara/Feature/Location.h

s9latimm committed:
Use ninja verbose

M	.travis.yml
M	CMakeLists.txt
M	include/vara/Feature/Location.h
M	lib/Feature/FeatureModelParser.cpp
M	unittests/Feature/Location.cpp

s9latimm committed:
Single job

M	.travis.yml

s9latimm committed:
Refactor

M	CMakeLists.txt
M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/FeatureModelParser.h
M	include/vara/Feature/Location.h
M	unittests/Feature/FeatureModelParser.cpp

s9latimm committed:
Experimental

M	CMakeLists.txt
M	include/vara/Feature/Location.h

s9latimm committed:
Fix lib

M	include/vara/Feature/FeatureModel.h
M	lib/Feature/CMakeLists.txt

s9latimm committed:
Enable targets

M	.travis.yml

s9latimm committed:
Use std

M	.travis.yml
M	include/vara/Feature/Location.h

s9latimm committed:
Disable tidy

M	.travis.yml

s9latimm committed:
Use experimental

M	include/vara/Feature/Location.h

s9latimm committed:
Enable llvm-10

M	.travis.yml

s9latimm committed:
Update travis

M	.travis.yml

s9latimm committed:
Update travis

M	.travis.yml

s9latimm committed:
Update travis

M	.travis.yml

s9latimm committed:
Fix tidy legacy

M	CMakeLists.txt

s9latimm committed:
Remove experimental

M	include/vara/Feature/Location.h

s9latimm committed:
Revert

M	include/vara/Feature/Location.h

s9latimm committed:
Only branches

M	.travis.yml

s9latimm committed:
Remove test

M	.gitignore
M	unittests/Feature/CMakeLists.txt
D	unittests/Feature/FeatureModelParser.cpp

s9latimm committed:
Merge branch 'f-Travis-Filesystem' into f-Location

M	.travis.yml
M	CMakeLists.txt
M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/FeatureModelParser.h
M	include/vara/Feature/Location.h
M	lib/Feature/CMakeLists.txt
M	lib/Feature/FeatureModelParser.cpp
M	unittests/Feature/Location.cpp

s9latimm committed:
Revert travis syntax

M	.travis.yml

s9latimm committed:
Switch include

M	.travis.yml
M	CMakeLists.txt
M	include/vara/Feature/Location.h
M	lib/Feature/CMakeLists.txt

vulder commented on include/vara/Feature/Location.h:
```diff
@@ -0,0 +1,69 @@
+
+#include <filesystem>
+namespace fs = std::filesystem;
+
```
```suggestion
```

vulder commented on include/vara/Feature/Location.h:
```diff
@@ -0,0 +1,69 @@
+namespace fs = std::experimental::filesystem;
+
+#else
+
```
```suggestion
```

vulder commented on include/vara/Feature/Location.h:
```diff
@@ -0,0 +1,69 @@
+
+#include <experimental/filesystem>
+namespace fs = std::experimental::filesystem;
+
```
```suggestion
```

vulder commented on include/vara/Feature/Location.h:
```diff
@@ -0,0 +1,69 @@
+
+#if defined(STD_EXPERIMENTAL_FILESYSTEM)
+#warning "using std::experimental::filesystem"
+
```
```suggestion
```
In the llvm style headers do not separate the parts inside a ifdef with extra lines

vulder commented on include/vara/Feature/Location.h:
```diff
@@ -0,0 +1,69 @@
+#define VARA_FEATURE_LOCATION_H
+
+#if defined(STD_EXPERIMENTAL_FILESYSTEM)
+#warning "using std::experimental::filesystem"
```
```suggestion
```
I don't think we need a warning in this case. It's not like the code is wrong or something, it's just that the user has an compiler that is not new "enough" to directly support this.

vulder commented on .travis.yml:
```diff
@@ -15,60 +15,60 @@ jobs:
             - llvm-9-dev
-            - clang-tidy-10
+            - clang++-9
+            - clang-tidy-9
```
```suggestion
            - clang-tidy-10
```
We should use the same tidy checker for all versions, otherwise, the error/warnings messages get out of sync, e.g., for false positives / false negatives that are fixed in newer versions.

vulder reviewed (changes requested):
Otherwise, this looks good :)

s9latimm committed:
Fix check

M	CMakeLists.txt
M	include/vara/Feature/Location.h
M	lib/Feature/CMakeLists.txt

s9latimm committed:
NOLINT

M	include/vara/Feature/Location.h

s9latimm committed:
Requested changes

M	.travis.yml
M	CMakeLists.txt
M	include/vara/Feature/Location.h
